// xfly/handlers/njshandler.js


var FS = require('fs');
var Path = require('path');

exports.processRequest = function(context){

	FS.readFile(context.request.physicalPath, context.applicationInstance.fileEncoding, function(e, data){
		if (e) {
			context.reportError(500, e);
			return;
		}

		context.response.contentEncoding = 'UTF-8';
		context.response.contentType = 'text/html; charset=UTF-8';
		
		var vmContext = initContext(context);
		
		var VM = require('vm');
		
		try{
			VM.runInNewContext(data, vmContext, context.request.physicalPath);
		} catch(e) {
			context.reportError(500, e);
		}
	});
	
};

function initContext(context){
	return {
		
		request: context.request,
		
		response: context.response,
		
		applicationInstance: context.applicationInstance,
		
		require: function (module) {
			if(module.indexOf('.') !== -1){
				module = Path.resolve(Path.dirname(context.request.physicalPath), module);
			}
			
			return require.call(this, module);
		},
		
		HttpUtility: require('../lib/httputility'),
		
		HttpCookie: require('../lib/httpcookie'),
		
		context: context,

		console: console,

		setTimeout: setTimeout,

		setInterval: setInterval
		
	};
	
}
