// utilskit/io.js
/**
 * @fileDescription Extended IO Utilities for node.js
 * @author xuld
 * @license MIT license
 */
 
var FS = require('fs');
var Path = require('path');
var Util = require('util');

function copyFile(srcFile, destFile){
	var BUF_LENGTH,
		BUF_LENGTH = 64 * 1024,
		buff = new Buffer(BUF_LENGTH),
		fdr = FS.openSync(srcFile, 'r'),
		fdw = FS.openSync(destFile, 'w'),
		bytesRead = 1,
		pos = 0;
	while(bytesRead > 0) {
		bytesRead = FS.readSync(fdr, buff, 0, BUF_LENGTH, pos);
		FS.writeSync(fdw, buff, 0, bytesRead);
		pos += bytesRead;
	}
	FS.closeSync(fdr);
	return FS.closeSync(fdw);
}

function copyDir(sourceDir, newDirLocation) {

    /*  Create the directory where all our junk is moving to; read the mode of the source directory and mirror it */
    var checkDir = FS.statSync(sourceDir);
    try {
        FS.mkdirSync(newDirLocation, checkDir.mode);
    } catch (e) {
        //if the directory already exists, that's okay
        if (e.code !== 'EEXIST') throw e;
    }

    var files = FS.readdirSync(sourceDir);

    for(var i = 0; i < files.length; i++) {
        var currFile = FS.lstatSync(sourceDir + "/" + files[i]);

        if(currFile.isDirectory()) {
            /*  recursion this thing right on back. */
            copyDir(sourceDir + "/" + files[i], newDirLocation + "/" + files[i]);
        } else if(currFile.isSymbolicLink()) {
            var symlinkFull = FS.readlinkSync(sourceDir + "/" + files[i]);
            FS.symlinkSync(symlinkFull, newDirLocation + "/" + files[i]);
        } else {
            /*  At this point, we've hit a file actually worth copying... so copy it on over. */
            copyFile(sourceDir + "/" + files[i], newDirLocation + "/" + files[i]);
        }
    }
}

function createDir (p, mode) {
    p = Path.resolve(p);

    try {
        FS.mkdirSync(p, mode);
    } catch (err0) {
        switch (err0.code) {
            case 'ENOENT' :
                createDir(Path.dirname(p), mode);
                createDir(p, mode);
                break;

            case 'EEXIST' :
                var stat;
                try {
                    stat = FS.statSync(p);
                } catch (err1) {
                    throw err0;
                }
                if (!stat.isDirectory()) throw err0;
                break;
            default :
                throw err0
                break;
        }
    }

};

function getFiles(path, basePath, files, directories){
	var rr = FS.readdirSync(path);
	
	path = path + Path.sep;
	for(var i = 0; i < rr.length; i++){
		var newPath = path + rr[i];
		var s = FS.statSync(newPath);
		
		if(s.isFile()){
			if(files) {
				files.push(basePath + rr[i]);
			}
		} else if(s.isDirectory()){
			if(directories) {
				directories.push(basePath + rr[i] + '/');
			}
			
			getFiles(newPath, basePath + rr[i] + '/', files, directories);
		}
	}
}

function cleanDir(path){
	var rr = FS.readdirSync(path);
	
	path = path + Path.sep;
	for(var i = 0; i < rr.length; i++){
		var newPath = path + rr[i];
		var s = FS.statSync(newPath);
		
		if(s.isFile()){
			FS.unlinkSync(newPath);
		} else if(s.isDirectory()){
			icleanDir(newPath);
		}
	}
}

var IO = {

	exists: FS.existsSync,
	
	existsFile: function(path) {
		return FS.existsSync(path) && FS.statSync(path).isFile();
	},
	
	existsDir: function(path) {
		return FS.existsSync(path) && FS.statSync(path).isDirectory();
	},
	
	ensureDir: function(path) {
		path = Path.dirname(path);
		if(!IO.exists(path))
			return createDir(path);
	},
	
	copyFile: function(srcFile, destFile, overwrite) {
		if(!IO.exists(srcFile))
			return false;
		
		if(IO.exists(destFile)){
			if(overwrite === false) {
				return false;
			} else {
				FS.unlinkSync(destFile);
			}
		} else {
			IO.ensureDir(destFile);
		}
		
		copyFile(srcFile, destFile);
		return true;
	},
	
	copyDir: function(src, dest) {
		if(!IO.exists(src))
			return false;
		
		copyDir(src, dest);
	},
	
	createDir: createDir,
	
	readFile: function(path, encoding) {
		if(IO.exists(path)) {
			encoding = encoding || "utf-8";
			var c = FS.readFileSync(path, encoding);
			if(/^utf\-?8/.test(encoding)){
				c = c.replace(/^\uFEFF/, '');
			}
			return c;
		}
		return '';
	},
	
	writeFile: function(path, content, encoding) {
		IO.ensureDir(path);
		
		FS.writeFileSync(path, content, encoding || "utf-8");
	},
	
	openWrite: function(path, options){
		IO.ensureDir(path);
		return FS.createWriteStream(path, options);
	},
	
	deleteFile: function(path) {
		if(IO.exists(path))
			FS.unlinkSync(path);
	},
	
	readDir: function(path){
		return IO.exists(path) ? FS.readdirSync(path) : [];
	},
	
	getFiles: function(path){
		if(!IO.exists(path)){
			return [];
		}
		
		var r = [];
		
		getFiles(path, "", r, null);
		
		return r;
	},
	
	getDirs: function(path){
		if(!IO.exists(path)){
			return [];
		}
		
		var r = [];
		
		getFiles(path, "", null, r);
		
		return r;
	},
	
	getDirAndFiles: function(path){
		if(!IO.exists(path)){
			return [];
		}
		
		var r = [];
		
		getFiles(path, "", r, r);
		
		return r;
	},
	
	cleanDir: function(path){
		if(IO.exists(path)) {
			cleanDir(path);
		}
	},
	
	deleteDir: function(path) {
		if(IO.exists(path)) {
			cleanDir(path);
			FS.unlinkSync(path);
		}
	}

};

module.exports = IO;
